### PR TITLE
fix: Update Firebase token when new comes (AR-2489)

### PR DESCRIPTION
**Pull request by borichellow**
# What's new in this PR?

new flow for registering Firebase token 

### Issues

Firebase token is not updated on our BE side if Firebase decides to create a new one (it's rare but possible case).

### Causes (Optional)

FB token when it comes was just saving into local DB and when user login we register that token for the user.

### Solutions

Add some flag into UserScope "shouldUpdateFirebaseToken"
On every `init` in `UserScope` check this `shouldUpdateFirebaseToken` flag if it's true - update the token for BE.
And everytime Firebase generate a new token we save it into local DB **and** set `shouldUpdateFirebaseToken` for all users to `true`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/notification/NotificationTokenRepository.kt [hunk 1]
```diff
@@ -6,22 +6,22 @@ import com.wire.kalium.logic.functional.map
 import com.wire.kalium.logic.wrapStorageRequest
 import com.wire.kalium.persistence.client.TokenStorage
 
-data class NotificationToken(val token: String, val transport: String)
+data class NotificationToken(val token: String, val transport: String, val senderId: String)
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/notification/NotificationTokenRepository.kt:9]: Can we specify what is `senderId`? Is it the client ID? Or User ID?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/user/UserRepository.kt [hunk 2]
```diff
@@ -62,6 +62,8 @@ internal interface UserRepository {
     suspend fun getAllKnownUsersNotInConversation(conversationId: ConversationId): Either<StorageFailure, List<OtherUser>>
     suspend fun getUsersFromTeam(teamId: TeamId): Either<StorageFailure, List<OtherUser>>
     suspend fun getTeamRecipients(teamId: TeamId): Either<CoreFailure, List<Recipient>>
+    suspend fun observeUpdateFirebaseTokenFlag(): Flow<Boolean>
+    suspend fun setUpdateFirebaseTokenFlag(shouldUpdate: Boolean): Either<StorageFailure, Unit>
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/user/UserRepository.kt:66]: This Repo already has too many responsibilities.
We could create a `PushTokenRepository` that has these two functions there.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/notificationToken/ObserveNotificationTokenUpdating.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import com.wire.kalium.logic.data.user.UserRepository
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.isRight
+import com.wire.kalium.network.api.user.pushToken.PushTokenBody
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.map
+
+internal interface ObserveNotificationTokenUpdating {
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/notificationToken/ObserveNotificationTokenUpdating.kt:14]: As this is not a UseCase, we can name it like a normal type and not like a function.

```suggestion
internal interface PushTokenUpdater
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/notificationToken/ObserveNotificationTokenUpdating.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    suspend operator fun invoke()
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/notificationToken/ObserveNotificationTokenUpdating.kt:15]: Perhaps

```suggestion
    suspend fun monitorTokenChanges()
```

This way, the utilisation becomes:

```kotlin
launch {
    val pushTokenUpdater = createPushTokenUpdater()
    pushTokenUpdater.monitorTokenChanges()
}
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/notificationToken/ObserveNotificationTokenUpdating.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                                senderId = notificationToken.senderId,
+                                client = clientId.value,
+                                token = notificationToken.token,
+                                transport = notificationToken.transport
+                            )
+                        )
+                    }
+            }
+            .collect { result ->
+                if (result.isRight()) userRepository.setUpdateFirebaseTokenFlag(false)
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/notificationToken/ObserveNotificationTokenUpdating.kt:43]: Really cool 😎

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt [hunk 6]
```diff
@@ -666,6 +675,8 @@ abstract class UserSessionScopeCommon internal constructor(
             slowSyncManager
 
             callRepository.updateOpenCallsToClosedStatus()
+
+            observeNotificationTokenUpdating()
         }
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:680]: Maybe we could launch this on its own job, so people don't forget that this operation is long suspending.

```suggestion
        }
        launch {
            observeNotificationTokenUpdating()
        }
```

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Really really cool 😎

Just a couple of tiny naming suggestions.

**Review by MohammedMokresh — APPROVED** [review]
LGTM

**Review by vitorhugods — APPROVED** [review]
🚀👩🏼‍🚀👨🏼‍🚀🧑🏼‍🚀

**Pull request merged**